cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -602,43 +603,109 @@ func (m *McmManager) retry(fn func(ctx context.Context) (bool, error), resourceT
 	}
 }
 
-// GetMachineDeploymentNodes returns the set of Nodes which belongs to the MachineDeployment.
-func (m *McmManager) GetMachineDeploymentNodes(machinedeployment *MachineDeployment) ([]string, error) {
-	md, err := m.machineDeploymentLister.MachineDeployments(m.namespace).Get(machinedeployment.Name)
-	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch MachineDeployment object %s, Error: %v", machinedeployment.Name, err)
-	}
+// GetInstancesForMachineDeployment returns list of cloudprovider.Instance for machines which belongs to the MachineDeployment.
+func (m *McmManager) GetInstancesForMachineDeployment(machinedeployment *MachineDeployment) ([]cloudprovider.Instance, error) {
+	var (
+		list     = []string{machinedeployment.Name}
+		selector = labels.NewSelector()
+		req, _   = labels.NewRequirement("name", selection.Equals, list)
+	)
 
-	machineList, err := m.machineLister.Machines(m.namespace).List(labels.Everything())
+	selector = selector.Add(*req)
+	machineList, err := m.machineLister.Machines(m.namespace).List(selector)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Machine objects %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Machine objects %v for machinedeployment %q", err, machinedeployment.Name)
 	}
 
 	nodeList, err := m.nodeLister.List(labels.Everything())
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Nodes %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Nodes %v", err)
 	}
 
-	var nodes []string
+	var instances []cloudprovider.Instance
 	// Bearing O(n2) complexity, assuming we will not have lot of nodes/machines, open for optimisations.
 	for _, machine := range machineList {
-		if strings.Contains(machine.Name, md.Name) {
-			var found bool
-			for _, node := range nodeList {
-				if machine.Labels["node"] == node.Name {
-					nodes = append(nodes, node.Spec.ProviderID)
-					found = true
-					break
+		instance := cloudprovider.Instance{}
+		var found bool
+		for _, node := range nodeList {
+			if machine.Labels["node"] == node.Name {
+				instance.Id = node.Spec.ProviderID
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No k8s node found - either the VM is up but has not registered yet or registered but node is NotReady or MCM is unable to fulfill the request to create VM.
+			// Report a special placeholder ID so that the autoscaler can track it as an unregistered node.
+			instance.Id = placeholderInstanceIDForMachineObj(machine.Name)
+			instance.Status = &cloudprovider.InstanceStatus{
+				State:     cloudprovider.InstanceCreating,
```

Comment:
Why is it assumed that the State will always be `cloudprovider.InstanceCreating`?

Reply:
I agree. Ideally `getErrorInfo(machine)` is called separately. Then if its `ErrorClass` matches `cloudprovider.OutOfResourcesErrorClass` (which already does the check for `LastOperation.Type`, only then the `State` should be assumed as `cloudprovider.InstanceCreating`.

Reply:
This decision doesn't have any implications on the flow, as autoscaler currently acts differently only for `Creating` machine with errors
But the decision only does mapping of `Creating` in CA to {"" , "Pending", "CrashloopBackoff"} of MCM.. 
also since other possible states are `Running`, `Deleting` or `no state` and in the eyes of autoscaler its still a `Creating` machine , thats why took this decision... 
I can revert it if you feel otherwise